protocol/client/messenger_v2.go

```diff
@@ -157,19 +162,55 @@ func (m *MessengerV2) Request(ctx context.Context, c Contact, options protocol.R
 	return m.proto.Request(ctx, options)
 }
 
+func (m *MessengerV2) requestHistories(ctx context.Context, histories []History, opts protocol.RequestOptions) error {
+	log.Printf("[messenger::RequestAll] requesting messages for chats %+v: from %d to %d\n", opts.Chats, opts.From, opts.To)
+	start := time.Now()
+	err := m.proto.Request(ctx, opts)
+	if err != nil {
+		return err
+	}
+	log.Printf("[messenger::RequestAll] requesting message for chats %+v finished. took %v\n", opts.Chats, time.Since(start))
+	for i := range histories {
+		histories[i].Synced = opts.To
+	}
+	err = m.db.UpdateHistories(histories)
+	return err
+}
+
 func (m *MessengerV2) RequestAll(ctx context.Context, newest bool) error {
-	contacts, err := m.db.Contacts()
+	// FIXME(dshulyak) if newest is false request 24h?
+	histories, err := m.db.Histories()
 	if err != nil {
 		return errors.Wrap(err, "error fetching contacts")
 	}
-	requestParams := protocol.DefaultRequestOptions()
-	for _, c := range contacts {
-		err = enhanceRequestOptions(c, &requestParams)
+	var (
+		now               = time.Now()
+		synced, notsynced = splitIntoSyncedNotSynced(histories)
+		errors            = make(chan error, 2)
+		wg                sync.WaitGroup
+	)
+	if len(synced) != 0 {
+		wg.Add(1)
+		go func() {
+			errors <- m.requestHistories(ctx, synced, syncedToOpts(synced, now))
+			wg.Done()
+		}()
+	}
+	if len(notsynced) != 0 {
+		wg.Add(1)
+		go func() {
+			errors <- m.requestHistories(ctx, notsynced, notsyncedToOpts(notsynced, now))
+			wg.Done()
+		}()
+	}
+	wg.Wait()
```

Comment:
What will happen if the counter for the working group is zero? Will it block or just return immediately here?

Reply:
it exits